src/redis/fixed_cache_impl.go

```diff
@@ -60,45 +67,67 @@ func (this *fixedRateLimitCacheImpl) DoLimit(
 
 		logger.Debugf("looking up cache key: %s", cacheKey.Key)
 
-		expirationSeconds := utils.UnitToDivider(limits[i].Limit.Unit)
-		if this.baseRateLimiter.ExpirationJitterMaxSeconds > 0 {
-			expirationSeconds += this.baseRateLimiter.JitterRand.Int63n(this.baseRateLimiter.ExpirationJitterMaxSeconds)
-		}
-
 		// Use the perSecondConn if it is not nil and the cacheKey represents a per second Limit.
 		if this.perSecondClient != nil && cacheKey.PerSecond {
-			if perSecondPipeline == nil {
-				perSecondPipeline = Pipeline{}
+			value, err := this.perSecondClient.GetValue(cacheKey.Key)
+			if err != nil {
+				logger.Error(err)
 			}
-			pipelineAppend(this.perSecondClient, &perSecondPipeline, cacheKey.Key, hitsAddend, &results[i], expirationSeconds)
+
+			results[i] = value
 		} else {
-			if pipeline == nil {
-				pipeline = Pipeline{}
+			value, err := this.client.GetValue(cacheKey.Key)
+			if err != nil {
+				logger.Error(err)
 			}
-			pipelineAppend(this.client, &pipeline, cacheKey.Key, hitsAddend, &results[i], expirationSeconds)
-		}
-	}
 
-	if pipeline != nil {
-		checkError(this.client.PipeDo(pipeline))
-	}
-	if perSecondPipeline != nil {
-		checkError(this.perSecondClient.PipeDo(perSecondPipeline))
+			results[i] = value
+		}
 	}
 
 	// Now fetch the pipeline.
 	responseDescriptorStatuses := make([]*pb.RateLimitResponse_DescriptorStatus,
 		len(request.Descriptors))
 	for i, cacheKey := range cacheKeys {
 
-		limitAfterIncrease := results[i]
-		limitBeforeIncrease := limitAfterIncrease - hitsAddend
+		limitBeforeIncrease := uint32(results[i])
+		limitAfterIncrease := limitBeforeIncrease + hitsAddend
 
 		limitInfo := limiter.NewRateLimitInfo(limits[i], limitBeforeIncrease, limitAfterIncrease, 0, 0)
 
 		responseDescriptorStatuses[i] = this.baseRateLimiter.GetResponseDescriptorStatus(cacheKey.Key,
 			limitInfo, isOverLimitWithLocalCache[i], hitsAddend)
 
+		if cacheKey.Key == "" || isOverLimitWithLocalCache[i] {
+			continue
+		}
+
+		expirationSeconds := utils.UnitToDivider(limits[i].Limit.Unit)
+		if this.expirationJitterMaxSeconds > 0 {
+			expirationSeconds += this.jitterRand.Int63n(this.expirationJitterMaxSeconds)
+		}
+
+		if this.perSecondClient != nil && cacheKey.PerSecond {
+			err := this.perSecondClient.IncrementValue(cacheKey.Key, uint64(hitsAddend))
+			if err != nil {
+				logger.Error(err)
+			}
+
+			err = this.perSecondClient.SetExpire(cacheKey.Key, uint64(expirationSeconds))
+			if err != nil {
+				logger.Error(err)
+			}
+		} else {
+			err := this.client.IncrementValue(cacheKey.Key, uint64(hitsAddend))
+			if err != nil {
+				logger.Error(err)
+			}
+
+			err = this.client.SetExpire(cacheKey.Key, uint64(expirationSeconds))
+			if err != nil {
+				logger.Error(err)
+			}
+		}
 	}
```

Comment:
this section looks like duplicated code

Reply:
yeah but it's not, it's using different client to execute

Reply:
perhaps you could extract it to a function that accepts a client?